### PR TITLE
Yaml metadata support

### PR DESCRIPTION
**Pull request by jkrautmacher**
This branch adds the `generate-metadata-yaml` script which is similar to the already existing `generate-version-h` script but adapted to be used with YAML files instead of C header files.

Furthermore two new environment variables are provided by the `git.sh` library:

- `__GIT_HASH_FULL__`
- `__GIT_COMMITDATE_UNIX__`

This extends the use cases in which UFW can generate metadata files to be included in builds.

Additionally an issue is fixed which led to not reproducible builds.

### CODE REVIEW

File: vcs-integration/git.sh [hunk 1]
```diff
@@ -63,7 +63,7 @@ _git_increment_ () {
 }
 
 _git_commitdate_ () {
-    REPLY="$(date -d @"$(git show -s --format=%ct --date=local $1)" '+%Y-%m-%d')"
+    REPLY="$(date -d @"$(git show -s --format=%ct $1)" '+%Y-%m-%d')"
```
Comment (ft) [inline vcs-integration/git.sh]: The `--date=local` (which is an alias for `--date=default-local`) has no impact with `%ct`, so there is no reproducibility problem with the way `git` is called here. Tough, I agree it is misleading to have the option in there. I cannot recall how this happened. Likely I was experimenting with sets of options until I found something consistent.

The problem is with using `date`. It can create a problem when not using `-u`, because we're reinterpreting the time-stamp produced by `git` — this is well spotted. But more significantly: `-d` for `date` is not portable **and** `git` can do all this on its own, without the call to `date`:

- `git show -s --date=format:%Y-%m-%d --format=%cd`

This is stable across different machines, and portable as long as `git` is installed.

Now in another commit you are arguing for dates to be yielded as though the `UTC` time-zone was active. I think that is not very clear cut an opinion: It can also be argued, that the current time for the person performing the act (authoring a change or committing this change to a repository) has valuable meaning. With time-zones close to UTC, this different is negligible (especially when only the day is stated); with further away zones, this is less so.

Now, it is possible to coerce `git` into calculating the day in Greenwich at this point in time as well:

- `TZ=UTC git show -s --date=format-local:%Y-%m-%d --format=%cd`

But I frankly don't see the point. I think this function should do thing I outlined before.

File: bin/generate-metadata-yaml [hunk 2]
```diff
@@ -0,0 +1,62 @@
+#!/bin/sh
+
+set -euf
```
Comment (ft) [inline bin/generate-metadata-yaml:3]: Not a fan of this set of options. I think `-e` can be confusing and people should better exit explicitly. `-u` is similarly obnoxious, with very idiomatic shell code nonetheless. And `-f` disables one of the more useful shell features. So yeah.

Also, the script uses a style that is quite significantly different from the other shell code in the project. And it uses some non-portable features (like `realpath`).

But that's all secondary: The way the parameter expansion is done here is a fair bit more limited than what the `version.h` generator allows. I think that generator should be modularised and its parts reused so the two generators show the same behaviour and have the same set of features.

File: vcs-integration/git.sh [hunk 3]
```diff
@@ -64,7 +64,7 @@ _git_increment_ () {
 }
 
 _git_commitdate_ () {
-    REPLY="$(date -d @"$(git show -s --format=%ct)" '+%Y-%m-%d')"
+    REPLY="$(date -ud @"$(git show -s --format=%ct)" '+%Y-%m-%d')"
```
Comment (ft) [inline vcs-integration/git.sh]: I addressed this elsewhere.

File: vcs-integration/git.sh [hunk 4]
```diff
@@ -7,6 +7,7 @@
 #
 #         __GIT_BRANCH__
 #         __GIT_COMMITDATE__
+#         __GIT_COMMITDATE_UNIX__
```
Comment (ft) [inline vcs-integration/git.sh:10]: This belongs to a different commit.

Sidenote: I've read elsewhere, that these `__GIT_*` variables, are `environment` variables: This is wrong: They are just variables set in the current shell. None of them are exported to the execution environment.

File: vcs-integration/git.sh [hunk 5]
```diff
@@ -168,8 +170,9 @@ git_populate () {
             __GIT_DIRTY__=0
         fi
 
-        _git_hash_
-        __GIT_HASH__="$REPLY"
+        _git_hash_full_
+        __GIT_HASH_FULL__="$REPLY"
+        __GIT_HASH__="$(echo "$REPLY" | cut -c1-12)"
```
Comment (ft) [inline vcs-integration/git.sh:175]: `echo` is a portability nightmare.

Also, upon reflection, we can get away with not even using `cut` at all. So that's a win:

```sh
_git_hash_ () {
    REPLY="$(git rev-parse --short=12 HEAD)"
}
```

…and define `_git_hash_full_` just like you did.

File: vcs-integration/git.sh [hunk 6]
```diff
@@ -63,7 +64,7 @@ _git_increment_ () {
 }
 
 _git_commitdate_ () {
-    REPLY="$(date -d @"$(git show -s --format=%ct $1)" '+%Y-%m-%d')"
+    REPLY="$(date -d @"$(git show -s --format=%ct)" '+%Y-%m-%d')"
```
Comment (ft) [inline vcs-integration/git.sh]: There is some inconsistency in the library. Probably, it would be best to use an argument for all these types of functions. Maybe do `committish="${1:-HEAD}"` (when it makes sense), which is what `git` often does internally.

It's true, that it is not used that way at the moment, but it improves future utility at virtually zero cost.

**Review by ft — COMMENTED** [review]
The problem with `date` is well spotted. I should have never done it that way.

The additional variables are also fair, albeit I'd do things differently as I've mentioned in my review notes.

The YAML generator has a reason for existing (mainly because of the boolean conversion), but the way it is implemented is lacking.

I will probably take a closer look at quickly fixing all my issues tomorrow. I couldn't motivate myself during the weekend, I'm afraid.

File: vcs-integration/git.sh (hunk 1 above)
Comment (jkrautmacher) [reply vcs-integration/git.sh]: Doing everything with `git` alone sounds like a good idea. I did not cover the portability issues of `date` so far.

Regarding the time zones: The library gets more flexibility when UTC is not enforced. The user can then freely choose a time zone and it works locally as expected.

My thoughts were very centered around a build process. In that case the output of the build process depends on the time zone of the machine on which the build is executed. If two people in different time zones try to get the same build result checksum from the same commit they will have a very hard time.

This is partially based on the issue that a date or even a full time stamp without time zone information is never precise. But here the API of the library exports exactly something like this. My thought was to pick one time zone which is always used to workaround the problem without an API change. UTC is then a bit more legitimate than all the other time zones if the user location is explicitly not considered.

But this is a difficult point and I see arguments for both sides. The explanation is just to document my thoughts maybe a bit better. Please choose what is the best option for `ufw` from your point of view.

File: bin/generate-metadata-yaml (hunk 2 above)
Comment (jkrautmacher) [reply bin/generate-metadata-yaml:3]: Oh these are actually my standard options since somebody passed me [this](https://sipb.mit.edu/doc/safe-shell/) link in a review. Personally I supported that because I:

- usually want my script to stop if something went wrong even when I forgot explicit error handling (`-e`)
- expect that a not explicitly set variable is most likely a typo or a similar bug (`-u`)
- did not use globbing and was afraid of unexpected magic (`-f`)

But thank you for your view on this topic. Will think more in detail about this next time.

For this code it is very simple and obvious: It should follow the conventions of the other scripts in the repository :)

**Comment by jkrautmacher** [comment]
Since the suggested changes are in master I am closing this here.
